### ISSUE TITLE
Slowdowns when using fixed_fraction_neon

### ISSUE THREAD

**Issue by georgeholt1** [bendudson/hermes-3#139]
I've noticed large slowdowns when running simulation with `fixed_fraction_neon`. There's an example below, which is adapted from the 1D-recycling example with added `fixed_fraction_carbon` and `fixed_fraction_neon`. I've run the same case with only `fixed_fraction_carbon` and saw no slowdown.

I'm running with Hermes-3 commit `0af0890` built with BOUT++ commit `17f728f`.

![2023-06-02_1_density_evolution](https://github.com/bendudson/hermes-3/assets/50923318/70748417-60ce-4084-9443-f4962d8265c8)

![2023-06-02_1_timer](https://github.com/bendudson/hermes-3/assets/50923318/a04c6e90-e5c5-4224-a70e-490341178f91)

Steady state is reached at around timestep 70, but at timestep 127 the code starts taking 1 or 2 orders of magnitude longer. Because this happens after reaching steady state it doesn't affect any results or physics, but it could be a problem in other cases where the slowdown starts happening before reaching steady state.

Input file:
```bash
nout = 400     # number of output time-steps
timestep = 5000.  # time between outputs

MXG = 0    # No guard cells needed in X



[mesh]

nx = 1
ny = 400    # Resolution along field-line
nz = 1

length = 25        # Length of the domain in meters
length_xpt = 12.5  # Length from midplane to X-point [m]

dymin = 0.1  # Minimum grid spacing near target, as fraction of average. Must be > 0 and < 1

dy = (length / ny) * (1 + (1-dymin)*(1-y/pi))   # Parallel grid spacing [m]

area_expansion = 1 # Expansion factor Area(target) / Area(midplane)
J = 1 + (mesh:area_expansion - 1) * H(y - mesh:y_xpt)*(y - mesh:y_xpt)/(2*pi - mesh:y_xpt)

# Calculate where the source ends in grid index
source = length_xpt / length
y_xpt = pi * ( 2 - dymin - sqrt( (2-dymin)^2 - 4*(1-dymin)*source ) ) / (1 - dymin)

ixseps1 = -1   # Branch-cut indices, specifying that
ixseps2 = -1   # the grid is in the SOL



[hermes]

# Evolve ion density, ion and electron pressure, then calculate force on ions due
# to electron pressure by using electron force balance.
components = (d+, d, e,
              sheath_boundary, collisions, recycling, reactions,
              electron_force_balance, neutral_parallel_diffusion,
              c, ne)

loadmetric = false
normalise_metric = true

Nnorm = 1e19
Bnorm = 1
Tnorm = 100



[solver]

type = beuler
use_precon=false  # False with beuler, True with cvode
snes_type = newtonls
ksp_type = gmres
max_nonlinear_iterations = 10

atol = 1e-7
rtol = 1e-5



[sheath_boundary]

lower_y = false
upper_y = true



[neutral_parallel_diffusion]

dneut = 10  # (B / Bpol)^2 in neutral diffusion terms



# Ions
# ----

[d+]  # Deuterium

type = (evolve_density, evolve_pressure, evolve_momentum,
        noflow_boundary, upstream_density_feedback)

noflow_lower_y = true
noflow_upper_y = false

charge = 1
AA = 2

density_upstream = 1e19
density_source_positive = false
density_controller_i = 5e-4
density_controller_p = 5e2

thermal_conduction = true

diagnose = true

recycle_as = d
recycle_multiplier = 1



[Nd+]

function = 1

source_shape = H(mesh:y_xpt - y) * 1e20



[Pd+]

function = 1

powerflux = 2e7
source = (powerflux*2/3 / (mesh:length_xpt)) * H(mesh:y_xpt - y)  # Input power as function of y



[NVd+]

function = 0



# Neutrals
# --------

[d]  # Deuterium

type = (evolve_density, evolve_pressure, evolve_momentum,
        noflow_boundary)

charge = 0
AA = 2

thermal_conduction = true

diagnose = true



[Nd]

function = 0.001



[Pd]

function = 0.0001



# Electrons
# ---------

[e]

type = quasineutral, evolve_pressure, zero_current, noflow_boundary

noflow_upper_y = false

charge = -1
AA = 1/1836

thermal_conduction = true

diagnose = true



[Pe]

function = `Pd+:function`

source = `Pd+:source`



# Carbon
# ------

[c]

type = fixed_fraction_carbon
fraction = 0.05  # 5% of electron density
diagnose = true  # To save radiation



# Neon
# ----

[ne]

type = fixed_fraction_neon
fraction = 0.05  # 5% of electron density
diagnose = true  # To save radiation



# Reactions and recycling
# -----------------------

[recycling]

species = d+



[reactions]

type = (
        d + e -> d+ + 2e,  # Deuterium ionisation
        d+ + e -> d,       # Deuterium recombination
        d + d+ -> d+ + d,  # Charge exchange
       )

diagnose = true
```

**Comment by mikekryjak** [comment]
Hi @georgeholt1, thanks for raising. Like we discussed last week, I have a strong suspicion that this is the same problem that I found with the argon curve: the curves are discontinuous. I found that when the simulation settles down to very near steady state, a part of the domain will inevitably land on the border between two segments of the curve and will snap back and forth, causing severe slowdowns.

I have now prepared Ar, Ne and N curves that should resolve this problem in https://github.com/bendudson/hermes-3/pull/129. Do you mind running the case on that branch and seeing if it resolves the issue?

**Comment by georgeholt1** [comment]
Looks like it's totally fixed with commit 83e43415a1f131a5530c6d6a2d92034776e8736b. Thanks!

![timer](https://github.com/bendudson/hermes-3/assets/50923318/71bf0868-6291-432b-a45a-38c50ea7c3f0)
